In [1]:
import pymongo

myclient = pymongo.MongoClient("mongodb://127.0.0.1:27017/")
mydb = myclient["Douyin"]
favorites = mydb.favorite

In [6]:
sample = favorites.find_one()
sample

{'_id': ObjectId('5de09fc31b9be34699ef2500'),
 'aweme_id': '6760968304870886669',
 'desc': '速度与激情🤟🏻',
 'create_time': 1574160608,
 'author': {'uid': '3003497551507358',
  'short_id': '2848172184',
  'nickname': '曾以深',
  'signature': '开心就好互相尊重\n😊❤️:T8T8080',
  'avatar_larger': {'uri': '2e32100062fff01a029a9',
   'url_list': ['https://p3-dy.byteimg.com/aweme/1080x1080/2e32100062fff01a029a9.jpeg',
    'https://p9-dy.byteimg.com/aweme/1080x1080/2e32100062fff01a029a9.jpeg',
    'https://p1-dy.byteimg.com/aweme/1080x1080/2e32100062fff01a029a9.jpeg'],
   'width': 720,
   'height': 720},
  'avatar_thumb': {'uri': '2e32100062fff01a029a9',
   'url_list': ['https://p3-dy.byteimg.com/aweme/100x100/2e32100062fff01a029a9.jpeg',
    'https://p9-dy.byteimg.com/aweme/100x100/2e32100062fff01a029a9.jpeg',
    'https://p1-dy.byteimg.com/aweme/100x100/2e32100062fff01a029a9.jpeg'],
   'width': 720,
   'height': 720},
  'avatar_medium': {'uri': '2e32100062fff01a029a9',
   'url_list': ['https://p3-dy.byteimg.

 'aweme_id': '6760968304870886669',
 'author_user_id': 3003497551507358,
 'desc': '速度与激情🤟🏻',
 'is_vr': False,
 'is_ads': False,
 'aweme_type': 0, 
 'sort_label': '',
 'video_text': [],
 
  'statistics': {'aweme_id': '6760968304870886669',
  'comment_count': 15397,
  'digg_count': 326639,
  'download_count': 6692,
  'play_count': 0,
  'share_count': 5705,
  'forward_count': 37,
  'lose_count': 0,
  'lose_comment_count': 0},

 'status': {'aweme_id': '6760968304870886669',
  'is_delete': False,
  'allow_share': True,
  'allow_comment': True,
  'is_private': False,
  'with_goods': False,
  'private_status': 0,
  'with_fusion_goods': False,
  'in_reviewing': False,
  'reviewed': 1,
  'self_see': False,
  'is_prohibited': False,
  'download_status': 0},


In [9]:
import pandas as pd

df_top = pd.DataFrame(list(favorites.find({}, {
    "_id" : 0,
    'aweme_id': 1,
    'is_vr': 1,
    'is_ads': 1,
    'aweme_type': 1, 
    'sort_label': 1
})))

In [10]:
df_statistics_data = []
for favorite in favorites.find({}, {
    "_id" : 0, 
    'statistics': 1
}):
    df_statistics_data.append(favorite['statistics'])


In [11]:
df_statistics = pd.DataFrame(df_statistics_data)

In [12]:
df_status_data = []
for favorite in favorites.find({}, {
    "_id" : 0, 
    'status': 1
}):
    df_status_data.append(favorite['status'])


In [13]:
df_status = pd.DataFrame(df_status_data)

In [38]:
df_top.head()

aweme_id  is_vr  is_ads  aweme_type sort_label
0  6760968304870886669  False   False           0           
1  6761363475487935757  False   False           0           
2  6760910924741201156  False   False           0           
3  6759091304006716684  False   False           0           
4  6764335221207747847  False   False           0

In [15]:
df_statistics.head()

aweme_id  comment_count  digg_count  download_count  play_count  \
0  6760968304870886669          15397      326639            6692           0   
1  6761363475487935757           1232       27844            2046           0   
2  6760910924741201156           5559      172021            2997           0   
3  6759091304006716684           4413      104796           29548           0   
4  6764335221207747847            100        1044              70           0   

   share_count  forward_count  lose_count  lose_comment_count  
0         5705             37           0                   0  
1         1721             71           0                   0  
2         1808             24           0                   0  
3        25585           2548           0                   0  
4           77              0           0                   0

In [16]:
df_status.head()

aweme_id  is_delete  allow_share  allow_comment  is_private  \
0  6760968304870886669      False         True           True       False   
1  6761363475487935757      False         True           True       False   
2  6760910924741201156      False         True           True       False   
3  6759091304006716684      False         True           True       False   
4  6764335221207747847      False         True           True       False   

   with_goods  private_status  with_fusion_goods  in_reviewing  reviewed  \
0       False               0              False         False         1   
1       False               0              False         False         1   
2       False               0              False         False         1   
3       False               0              False         False         1   
4       False               0              False         False         1   

   self_see  is_prohibited  download_status  
0     False          False                0  
1     False          False                0  
2     False          False                0  
3     False          False                0  
4     False          False                0

In [17]:
df_is_vr = df_top[df_top['is_vr'] == True]

In [37]:
df_is_vr.head()

Empty DataFrame
Columns: [aweme_id, is_vr, is_ads, aweme_type, sort_label]
Index: []

In [19]:
df_is_ads = df_top[df_top['is_ads'] == True]
df_is_ads.head()

aweme_id  is_vr  is_ads  aweme_type sort_label
163150  6622829853831859459  False    True           0           
246482  6621809544018267405  False    True           0           
246979  6620337357948194061  False    True           0           
250229  6617717586731732228  False    True           0           
250578  6618842475215719693  False    True           0

In [22]:
ids_ads = list(df_is_ads['aweme_id'])

In [29]:
select = []

for index, record in df_statistics.iterrows():
    if record['aweme_id'] in ids_ads:
        select.append(True)
    else:
        select.append(False)
    

In [30]:
df_statistics_ads = df_statistics[select]

In [36]:
df_statistics_ads.head()

aweme_id  comment_count  digg_count  download_count  \
163150  6622829853831859459           1508       72741            1876   
246482  6621809544018267405           2808       95262            1329   
246979  6620337357948194061           2279      132595            1190   
250229  6617717586731732228           1789       60370             488   
250578  6618842475215719693           2634       69150             307   

        play_count  share_count  forward_count  lose_count  lose_comment_count  
163150           0         1124              1           0                   0  
246482           0          833              5           0                   0  
246979           0         1012              2           0                   0  
250229           0          502              2           0                   0  
250578           0          232              7           0                   0

In [ ]:
users = mydb.users
df_user_data = []
counter = 0
for user in users.find():
    if counter % 100 == 0:
        print('process: ', counter)
    if counter >= 10000:
        break
    df_user_data.append(user['user'])
    counter += 1

In [42]:

df_user = pd.DataFrame(df_user_data)
df_user_reduce = df_user[['uid', 'nickname', 'is_star', 'gender', 'birthday', 'follower_count', 'following_count', 'total_favorited', 'dongtai_count',
       'with_shop_entry', 'commerce_user_level', 'type_label']]

In [43]:
df_user_reduce.head()

uid    nickname is_star  gender    birthday  follower_count  \
0  1270684175830861        娱乐工兵   False     0.0  1994-01-01        131392.0   
1       96146561821      萌叔大卫老师   False     1.0  1982-01-01       1292976.0   
2       57178761731        过气网猫   False     1.0  2019-01-01        746061.0   
3       85546771621  GRIT国际义工旅行   False     2.0  2014-01-01        146219.0   
4      105516809810         俩不正   False     0.0  1997-01-01        579119.0   

   following_count  total_favorited  dongtai_count with_shop_entry  \
0             64.0        4670964.0          190.0           False   
1              5.0        8918406.0          144.0           False   
2            137.0       21048330.0          148.0           False   
3              4.0        1684085.0          103.0           False   
4             18.0        4294064.0           60.0           False   

   commerce_user_level  type_label  
0                  0.0         NaN  
1                  0.0         NaN  
2                  0.0         NaN  
3                  1.0         NaN  
4                  0.0         NaN

In [44]:
df_author_id = pd.DataFrame(list(favorites.find({}, {
    "_id" : 0,
    'author_user_id': 1
})))

In [45]:
df_author_id_ads = df_author_id[select]

In [46]:
user_ids_ads = list(df_author_id_ads['author_user_id'])

In [48]:
user_ids_ads

[93247169373,
 95992785087,
 95992785087,
 95992785087,
 95992785087,
 95992785087,
 95992785087,
 95992785087,
 95992785087,
 95992785087,
 73255678233,
 73255678233]

In [50]:
select_user_is_ads = []

for index, record in df_user_reduce.iterrows():
    if record['uid'] in user_ids_ads:
        select_user_is_ads.append(True)
    else:
        select_user_is_ads.append(False)

In [53]:
df_user_reduce_ads = df_user_reduce[select_user_is_ads]

In [54]:
df_user_reduce_ads

Empty DataFrame
Columns: [uid, nickname, is_star, gender, birthday, follower_count, following_count, total_favorited, dongtai_count, with_shop_entry, commerce_user_level, type_label]
Index: []

In [111]:
mydb.list_collection_names(session=None)

['follow', 'favorite', 'comment', 'users', 'logging']

In [57]:
follows = mydb.follow

In [61]:
sample = follows.find_one()

In [112]:
rfollow = mydb.rfollow

In [64]:
sample['fromuserid']

'92551583843'

In [68]:
following_pairs = []
for following in sample['followings']:
    pair = {}
    pair['followed'] = following['uid']
    pair['follower'] = sample['fromuserid']
    following_pairs.append(pair)

In [113]:
for record in follows.find():
    following_pairs = []
    for following in record['followings']:
        pair = {}
        pair['followed'] = following['uid']
        pair['follower'] = record['fromuserid']
        following_pairs.append(pair)
    if len(following_pairs) > 0:
        result = rfollow.insert_many(following_pairs)

In [110]:
rfollow.drop()

In [101]:
rfollow.count_documents({})

44995

In [102]:
follows.count_documents({})

3277

In [125]:
aggs = [
    {
        "$group":{
            "_id": "$followed",
            "count": {"$sum":1},
            'followers': {'$addToSet': '$follower'}
        }
    }
]

result = rfollow.aggregate(pipeline=aggs)
count = 0

for i in result:
    print(i)
    print(len(i['followers']))
    count += i['count']

{'_id': '111185346649', 'count': 7, 'followers': ['80137162334']}
1
{'_id': '102566271198', 'count': 19, 'followers': ['76906558048']}
1
{'_id': '70317966616', 'count': 11, 'followers': ['103699672143']}
1
{'_id': '99997427874', 'count': 9, 'followers': ['2000727977240206']}
1
{'_id': '101793954777', 'count': 36, 'followers': ['3003497551507358']}
1
{'_id': '110730458569', 'count': 24, 'followers': ['97833534794', '2000727977240206']}
2
{'_id': '58804156452', 'count': 19, 'followers': ['52535828443']}
1
{'_id': '63622535338', 'count': 16, 'followers': ['109825363535']}
1
{'_id': '111009855989', 'count': 6, 'followers': ['80423805968']}
1
{'_id': '110817817474', 'count': 6, 'followers': ['103699672143']}
1
{'_id': '98272308412', 'count': 19, 'followers': ['96548896937', '21612781648173']}
2
{'_id': '70588344795', 'count': 19, 'followers': ['52535828443']}
1
{'_id': '108364642401', 'count': 3, 'followers': ['89872572780']}
1
{'_id': '109748114692', 'count': 17, 'followers': ['63570722628

{'_id': '85897230831', 'count': 18, 'followers': ['108308788390']}
1
{'_id': '99433781178', 'count': 10, 'followers': ['108308788390']}
1
{'_id': '2238223396386808', 'count': 20, 'followers': ['80423805968']}
1
{'_id': '60483067677', 'count': 16, 'followers': ['76199546279']}
1
{'_id': '106856706239', 'count': 18, 'followers': ['62675992539']}
1
{'_id': '4111797845492587', 'count': 1, 'followers': ['68265819499']}
1
{'_id': '55082577887', 'count': 10, 'followers': ['102466425954']}
1
{'_id': '82480311983', 'count': 22, 'followers': ['73439410307']}
1
{'_id': '57366759152', 'count': 65, 'followers': ['4028913685771']}
1
{'_id': '105391953670', 'count': 1, 'followers': ['101381722841']}
1
{'_id': '57259595825', 'count': 71, 'followers': ['57451550285']}
1
{'_id': '102206634833', 'count': 20, 'followers': ['93949062144']}
1
{'_id': '55535498870', 'count': 20, 'followers': ['95591948189']}
1
{'_id': '75919570331', 'count': 11, 'followers': ['80423805968']}
1
{'_id': '99329466460', 'count':

1
{'_id': '64690192235', 'count': 76, 'followers': ['61841769622']}
1
{'_id': '104332721863', 'count': 20, 'followers': ['96095711727']}
1
{'_id': '60282758675', 'count': 35, 'followers': ['100796096592328']}
1
{'_id': '52850561357', 'count': 12, 'followers': ['72449252503']}
1
{'_id': '75467875213', 'count': 20, 'followers': ['89872572780']}
1
{'_id': '57813370462', 'count': 5, 'followers': ['102347288907']}
1
{'_id': '109921194550', 'count': 4, 'followers': ['103699672143']}
1
{'_id': '82135212477', 'count': 19, 'followers': ['64551290646']}
1
{'_id': '109820860481', 'count': 3, 'followers': ['72798335878']}
1
{'_id': '66169543999', 'count': 19, 'followers': ['62591278384']}
1
{'_id': '62409572552', 'count': 20, 'followers': ['62675992539']}
1
{'_id': '92741700547', 'count': 13, 'followers': ['61363195725']}
1
{'_id': '59341249271', 'count': 20, 'followers': ['73439410307']}
1
{'_id': '3452090463554220', 'count': 10, 'followers': ['96095711727']}
1
{'_id': '57324174293', 'count': 20,

In [107]:
match = {'$match':{'followed':'111185346649'}}
result = rfollow.aggregate([match])

In [118]:
for i in result:
    print(len(i['followers']))

In [126]:
count

44995

In [131]:
for _followed in rfollow.distinct('followed'):
    for _follower in rfollow.distinct('follower'):
        # 删除所有id相同的数据
        result = rfollow.delete_many({'followed': _followed, 'follower': _follower})
        # 把刚刚复制的数据加入一条到数据库
        rfollow.insert_one({'followed': _followed, 'follower': _follower})

KeyboardInterrupt: 